From aba182adda97a02e3cbbdcc3061d91f0037b73ae Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 10 Jun 2020 08:21:07 +0300
Subject: [PATCH] [SW-9644] preboot: add debug mode framework

Preboot needs debug mode capabilities to skip
hardware initialization based on boot strap pin
settings. e.g. i2c slave, pci module.

Change-Id: Id6e9aeb47d4218f3c98925d9586772968ab1338c
---

diff --git a/gaudi/common.c b/gaudi/common.c
index 90cd82f..a31f773 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -27,6 +27,30 @@
 #include "common.h"
 
 #define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
+
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS);
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = BIT(19); /* Dump Select[11] */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(18); /* Dump Select[10] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(17); /* Dump Select[9] */
+		break;
+	default:
+		preboot_puts_hex("Unknown H/W function ", func, false);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
 static void preboot_init_pcie_phy_fw_update(void)
 {
 	/* Configure FW update size */
diff --git a/goya/common.c b/goya/common.c
index dc783f4..5e9c525 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -26,6 +26,29 @@
 
 #define BTL_COUNTER_MAX		BIT(22)
 
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS);
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = BIT(21); /* Dump Select[8] */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(20); /* Dump Select[7] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(19); /* Dump Select[6] */
+		break;
+	default:
+		preboot_puts_hex("Unknown H/W function ", func, false);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
 static int preboot_wait_for_bit(unsigned int reg, unsigned long timeout_ns)
 {
 	unsigned int val;
diff --git a/goya2/common.c b/goya2/common.c
index 07cc248..def71ec 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -28,6 +28,33 @@
 #include "common.h"
 
 #define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
+
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_DBG_CARD:
+		strap_mask = BIT(11); /* Dump Select[3] */
+		break;
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = BIT(10); /* Dump Select[2] */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(8); /* Dump Select[0] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(9); /* Dump Select[1] */
+		break;
+	default:
+		preboot_puts_hex("Unknown H/W function ", func, false);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
 static void preboot_init_pcie_phy_fw_update(void)
 {
 	/* Configure FW update size */
diff --git a/goya2_arc_gnu/common.c b/goya2_arc_gnu/common.c
index 6db9cfe..ad197d4 100644
--- a/goya2_arc_gnu/common.c
+++ b/goya2_arc_gnu/common.c
@@ -32,6 +32,32 @@
 
 #define FW_WORD_NUM		0x2200	/* FW_SIZE / 2 */
 
+bool hl_is_debug_mode(enum hl_debug_hw_func func)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
+	u32 strap_mask;
+
+	switch (func) {
+	case HL_DEBUG_HW_DBG_CARD:
+		strap_mask = BIT(11); /* Dump Select[3] */
+		break;
+	case HL_DEBUG_HW_PCIE:
+		strap_mask = BIT(10); /* Dump Select[2] */
+		break;
+	case HL_DEBUG_HW_DRAM:
+		strap_mask = BIT(8); /* Dump Select[0] */
+		break;
+	case HL_DEBUG_HW_I2C:
+		strap_mask = BIT(9); /* Dump Select[1] */
+		break;
+	default:
+		preboot_puts_hex("Unknown H/W function ", func, false);
+		return 0;
+	}
+
+	return boot_strap & strap_mask;
+}
+
 static void preboot_init_pcie_phy_fw_update(void)
 {
 	/* Configure FW update size */
diff --git a/include/preboot.h b/include/preboot.h
index 79c621a..857ed13 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -37,6 +37,7 @@
 
 #ifndef __ASSEMBLY__
 
+#include <stdbool.h>
 #include "auto_generated.h"
 
 #define VERSION_MAX_LEN		128
@@ -60,6 +61,13 @@
 	HL_PRSTN_RST = 2,
 	HL_WD_RST = 4,
 	HL_SW_ALL_RST = 8,
+};
+
+enum hl_debug_hw_func {
+	HL_DEBUG_HW_PCIE,
+	HL_DEBUG_HW_DRAM,
+	HL_DEBUG_HW_I2C,
+	HL_DEBUG_HW_DBG_CARD,
 };
 
 static inline u8 regmap_readb(u32 offset)
@@ -91,6 +99,7 @@
 u32 hl_spi_baudr_get(void);
 void hl_plat_init(void);
 void hl_ts_init(void);
+bool hl_is_debug_mode(enum hl_debug_hw_func func);
 
 void preboot_pcie_wait_for_link(void);
 void preboot_delay(void);
